[FLINK-13676][ml] Add class of Vector to Columns mapper

xuyang1706 opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](https://flink.apache.org/contributing/contribute-code.html#open-a-pull-request).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

VectorToColumnsMapper maps one vector to many column objects.  The number of columns equals to vector size.  


## Brief change log

- *Add VectorToColumnsMapper for mapping Op.*
- *Add VectorToColumnsParams for parameters.*
- *Add VectorToColumnsMapperTest for unit test.*
 

## Verifying this change
This change added tests and can be verified as follows:

- run test case pass

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

walterddr commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/dataproc/vector/VectorToColumnsMapper.java:
```diff
@@ -0,0 +1,104 @@
+ * under the License.
+ */
+
+package org.apache.flink.ml.common.dataproc.vector;
```
I was wondering if this is the right package to put this utility function in, maybe:
`org.apache.flink.ml.common.utils.dataproc` makes more sense?

This is especially confusing since we already have `org.apache.flink.ml.common.mapper` base class.

qiuxiafei replied:
This is not a utility class, it's a data preprocess mapper. So I think it shouldn't be in a util package.

walterddr replied:
I see. I am assuming this pre-processing is similar to, say feature normalization, or some specific pre-processing algorithm such as: [Word Embedding algorithm](https://en.wikipedia.org/wiki/Word_embedding)?

qiuxiafei replied:
yes, but this is a pretty simple one.

walterddr replied:
in this case I think the location is spot on :-) Thanks

walterddr commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/dataproc/vector/VectorToColumnsMapper.java:
```diff
@@ -0,0 +1,104 @@
+				if (indices[i] < colSize) {
+					result.setField(indices[i], values[i]);
+				} else {
+					break;
```
```
int i = 0;
while (indices[i] < colSize) {
  result.setField(indices[i], values[i]);
}
```
also, are we always assume users' sparse vector indices larger than colSize can be ignored?

qiuxiafei replied:
Yes, this mapper flattens the `idx`th vector column in the input row to the output row. Maybe some warning log if truncated? But I just worry there maybe too much such logs.

walterddr replied:
hmm. that's probably right. I think explaining this in the JavaDoc might be suffice.

walterddr commented on flink-ml-parent/flink-ml-lib/src/test/java/org/apache/flink/ml/common/dataproc/vector/VectorToColumnsMapperTest.java:
```diff
@@ -0,0 +1,101 @@
+ */
+public class VectorToColumnsMapperTest {
+	@Test
+	public void test1() throws Exception {
```
needs more informative test naming.

walterddr commented on flink-ml-parent/flink-ml-lib/src/test/java/org/apache/flink/ml/common/dataproc/vector/VectorToColumnsMapperTest.java:
```diff
@@ -0,0 +1,101 @@
+	}
+
+	@Test
+	public void test2() throws Exception {
```
ditto

walterddr commented on flink-ml-parent/flink-ml-lib/src/test/java/org/apache/flink/ml/common/dataproc/vector/VectorToColumnsMapperTest.java:
```diff
@@ -0,0 +1,101 @@
+	}
+
+	@Test
+	public void test3() throws Exception {
```
ditto

xuyang1706 replied:
Thanks, renamed the test case.

walterddr commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/dataproc/vector/VectorToColumnsMapper.java:
```diff
@@ -0,0 +1,104 @@
+import org.apache.flink.types.Row;
+
+/**
+ * This mapper maps vector to table columns.
```
I think if this is a particular pre-processing algorithm/method, it needs better Javadoc.

xuyang1706 replied:
Thanks, refined the JavaDoc

xuyang1706 commented:
> @xuyang1706 overall it looks good. I changed the `RowCollector` test bug in [walterddr@2cd2906](https://github.com/walterddr/flink/commit/2cd29067856ed2372a6186bfe03cdeb26e54dcb2) to get Travis pass. I just left minor comments. please kindly take a look.

Thanks @walterddr for your comments and discussion. I have refined the code, renamed the test cases and added more JavaDoc.

ex00 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/operator/common/dataproc/vector/VectorToColumnsMapper.java:
```diff
@@ -0,0 +1,112 @@
+	private int idx;
+	private OutputColsHelper outputColsHelper;
+
+	public VectorToColumnsMapper(TableSchema dataSchema, Params params) {
```
Could you let a java doc with description which parameters should set successful creating new object, please?

walterddr commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/operator/common/dataproc/vector/VectorToColumnsMapper.java:
```diff
@@ -0,0 +1,112 @@
+import java.util.Arrays;
+
+/**
+ * This mapper maps vector to table columns, and the table is created with the first
```
```suggestion
 * This is a data preprocessing function that transforms {@link Vector}s into {@link Table} columns.
 *
 * <p>Table is created with the first colSize value of the vector.
 * 
 * <p>For sparse vector without given size, it will be treated as vector with infinite size.
 * ...
```

walterddr commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/operator/common/dataproc/vector/VectorToColumnsMapper.java:
```diff
@@ -0,0 +1,112 @@
+		TypeInformation[] types = new TypeInformation[colSize];
+		Arrays.fill(types, Types.DOUBLE);
+		this.outputColsHelper = new OutputColsHelper(dataSchema, outputColNames, types,
+			this.params.get(VectorToColumnsParams.RESERVED_COLS));
```
Let's refactor this part out as a private static function: `constructOutputColsHelper` ? I can see @ex00 's concerns that the constructor is way too complex. In principle the constructor should be simpler and easy to understand. One example is to have this as:
```
public VectorToColumnsMapper(TableSchema dataSchema, Params params) {
  this(dataSchema, params, constructOutputColsHelper(dataSchema, params));
}

public VectorToColumnsMapper(TableSchema dataSchema, Params params, OutputColsHelper outputColsHelper) {
  super(dataSchema, params);
  this.outputColsHelper = outputColsHelper;
}
```

walterddr commented on flink-ml-parent/flink-ml-lib/src/test/java/org/apache/flink/ml/operator/common/dataproc/vector/VectorToColumnsMapperTest.java:
```diff
@@ -0,0 +1,109 @@
+/**
+ * Unit test for VectorToColumnsMapper.
+ */
+public class VectorToColumnsMapperTest {
```
Test case looks good! thanks for refining @xuyang1706
